### ISSUE TITLE
example hungs

### ISSUE THREAD

**Issue by kostya** [stefanwille/crystal-redis#2]
such example hungs infinitely:
```ruby
require "redis"

N = 1_000_000
redis = Redis.new
redis.pipelined do |pipeline|
  N.times do |i|
    pipeline.set("foo", "bar")
  end
end

```

strace shows this: `epoll_wait(3, `

if i change N to 1_000_00 all work.

```
cat .deps.lock 
{
  "redis": "17d83192b22f718ff1e9b5c96608db688255fdbc"
}

```

```
Crystal 0.7.2 [3034ebb] (Tue May 26 16:40:34 UTC 2015)
```

Ubuntu 14.04.1 LTS

**Comment by stefanwille** [comment]
Hi Kostya, 

thanks for the bug report!

I tried your code on my Mac and couldn't reproduce the issue. Which operating system are you on?

**Comment by stefanwille** [comment]
oops, just saw that you are on Ubuntu 14.04, never mind

**Comment by stefanwille** [comment]
I tried the example on Ubuntu 14.04.1 LTS / 64 bit inside a VMware fusion virtual machine - works fine.
Which version of the Redis server are you using?

**Comment by kostya** [comment]
on Ubuntu 12.04.1 LTS
`N = 39499` is ok, `N = 39500` hungs.
Redis server v=2.8.17

**Comment by stefanwille** [comment]
Hi Kostya,

I have tested on Ubuntu 12.04.5 and Redis 2.8.17, and I could reproduce the issue. I couldn't figure out what the issue is. 

As a workaround, I changed pipeline handling such that big pipelines effectively get split into multiple smaller ones. This works for me. Can you try the latest version please?

**Comment by kostya** [comment]
thanks, seems fixed

**Comment by asterite** [comment]
I wonder what the real problem is... if it's a bug in Crystal (either the std or the compiler) we shouldn't let this opportunity pass to fix a serious bug :-)

**Comment by stefanwille** [comment]
@asterite I agree, and also I would rather not have the workaround in the Crystal-Redis code.

I made some progress: I extracted the code that causes the lockup into a single Crystal source file. This makes it easier to experiment. Here is is:

```Crystal
require "socket"

socket = TCPSocket.new("localhost", 6379)
io = BufferedIO.new(socket)

N = 1_000_000

puts "Sending #{N} commands"
N.times do |i|
  if i % 1000 == 0
    puts "  #{i}"
  end
  io << "*3\r\n"
  io << "$3\r\nSET\r\n"
  io << "$3\r\nfoo\r\n"
  io << "$3\r\nbar\r\n"
end

puts "Flushing"
io.flush

puts "Receiving #{N} responses"
N.times do |i|
  line = io.gets
  if line != "+OK\r\n"
    raise "Got #{line}, expected +OK"
  end
end
```

This triggers the issue reliably on Ubuntu 12.04 and 14.04 (not on Mac OS) and Redis 2.8 / 3.0.
The program always hangs in the first loop, while sending the commands.

And if I replace the `BufferedIO` object with my socket object like so:

```Crystal
# io = BufferedIO.new(socket)
io = socket
```

then the program doesn't hang. It looks to me like a problem in the BufferedIO implementation. 

Do you have an idea on how I can dig deeper?

**Comment by asterite** [comment]
@stefanwille Nice reduction :-)

I was able to reproduce it, but it's not BufferedIO's fault. This also hungs:

```crystal
require "socket"

socket = TCPSocket.new("localhost", 6379)
io = socket

N = 1_000_000

puts "Sending #{N} commands"
N.times do |i|
  if i % 1000 == 0
    puts "  #{i}"
  end
  io << "*3\r\n$3\r\nSET\r\n$3\r\nfoo\r\n$3\r\nbar\r\n"
end

puts "Flushing"
io.flush
```

If I change Socket to be blocking (right now it's non-blocking) it doesn't hang, so it must be something related to that... I need to investigate further, but @waj is probably the one that will surely know what could be the problem.

**Comment by waj** [comment]
I think I fixed this with the latest commit (manastech/crystal@04fb587). Redis starts sending responses way before they are read by the client and that trigger events with both the "read" and "write" flags set.

@kostya, @stefanwille could you please confirm if this fixes the issue in your case?

**Comment by stefanwille** [comment]
@waj @asterite yes, this fixes the issue - tested on Ubuntu 14.04.
